### PR TITLE
Added mouse support to the stingray plugin.

### PR DESCRIPTION
**Pull request by thynnmas**
This includes functions in the debuginator itself to find and select
the closest item to a given height-from-root.

### CODE REVIEW

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp [hunk 1]
```diff
@@ -292,6 +295,8 @@ void handle_default_input(TheDebuginator* debuginator, unsigned devices) {
 		return;
 	}
 
+	float default_focus_height = 0.3f;
```
Comment (Srekel) [inline plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: Is this needed? Why not use the default in the debuginator?

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp [hunk 2]
```diff
@@ -376,6 +383,35 @@ void handle_default_input(TheDebuginator* debuginator, unsigned devices) {
...
+		input_wrapper.controller = mouse;
+
+		if (api->any_released(input_wrapper.controller) != UINT_MAX) {
+			input_wrapper.time_since_pressed = -1;
+		}
+
+
+		if (debuginator_is_open(debuginator)) {
+			CApiVector3 scroll = input_wrapper.axis("wheel");
+			if (scroll.y != 0) {
```
Comment (Srekel) [inline plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: Sure you don't want to check if the mouse is inside the menu rect?

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp [hunk 3]
```diff
@@ -376,6 +383,35 @@ void handle_default_input(TheDebuginator* debuginator, unsigned devices) {
...
+				debuginator->focus_height += scroll.y * 0.05f;
```
Comment (Srekel) [inline plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: I'm not sure about this solution.. The idea with focus_height is to say "I want the hot item to be at 30% from the top of the screen". Modifying it based on scrolling, presumably getting it beyond [0,1] range feels a bit strange..

Wouldn't it be more reasonable to change the hot item? Then everything would be taken care of "automatically".

File: the_debuginator.h [hunk 4]
```diff
@@ -2377,6 +2377,104 @@ void debuginator_move_to_parent(TheDebuginator* debuginator) {
...
+						if (descendent_distance >= debuginator->item_height || descendent_distance <= last_distance) {
+							descendent->leaf.hot_index = -1; // Mark that we don't want the active one
+						}
+					}
+					return descendent;
+				}
+				return last_child;
+			} else if (last_child->leaf.is_expanded) {
+				int descendent_distance = last_distance;
+				if (last_child->leaf.is_expanded) {
```
Comment (Srekel) [inline the_debuginator.h]: This if is the same as the if two lines up, line 2410.

**Review by Srekel — CHANGES_REQUESTED** [review]

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp (hunk 2 above)
Comment (thynnmas) [reply plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: We catch all input, which is why this felt natural to us (so you could just scroll without paying attention to where the mouse is if you're not using it to play). However I agree that in general you should not take away control when it is open, so it _should_ check for the rect.

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp (hunk 1 above)
Comment (thynnmas) [reply plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: I must have missed that there was a default (I only spotted the used focus_height which was constant, but I'm misusing that for scroll now so it changes over time).

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp (hunk 3 above)
Comment (thynnmas) [reply plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: The usage request was "I want to be able to scroll, then if I don't click something and press a up I should go one up from the old selection, like a text-editor or file explorer would", which is why I'm doing this. It feels weird that the focus_height is outside [0,1] range and we're focusing the view around a hot item that's not on screen; probably rather want a view-offset and let the focus height be constant.

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp (hunk 1 above)
Comment (Srekel) [reply plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: There's the one in the default config thing. It should really be added to DebuginatorPluginCreateContext and then you can store the app's wanted default value in this function similar to how other config variables are done:

TheDebuginator* create_debuginator(const char* id, DebuginatorPluginCreateContext* context) {

File: plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp (hunk 3 above)
Comment (Srekel) [reply plugins/stingray/the_debuginator_plugin/c_api_the_debuginator.cpp]: Right, ok that makes sense. :)

If it's ok with you, I'll accept this after you fix the other issues, but there are some things I think I would like to change with how this is done. I want to move some of the functionality to The Debuginator. I don't want each app to have to reinvent how to do mouse-input.

**Review by Srekel — CHANGES_REQUESTED** [review]
Since I've now added a mouse support API, please use that instead. Let me know if you have any issues with it. You can check the SDL demo for examples of how to use it :)

**Comment by Srekel** [comment]
I'm going to close this, but feel free to open a new one after you've updated to the new input API, if you needed to make any changes.  :)

**Pull request closed without merging**